tools/stern/src/Stern/Intra.hs

```diff
@@ -376,29 +375,27 @@ setLegalholdStatus tid status = do
          . paths ["/i/teams", toByteString' tid, "features", "legalhold"]
          . lbytes (encode $ toRequestBody status)
          . contentJson
-         . expect2xx
          )
   where
-    toRequestBody False = LegalHoldTeamConfig LegalHoldDisabled
-    toRequestBody True  = LegalHoldTeamConfig LegalHoldEnabled
+    toRequestBody SetLegalHoldDisabled = LegalHoldTeamConfig LegalHoldDisabled
+    toRequestBody SetLegalHoldEnabled  = LegalHoldTeamConfig LegalHoldEnabled
 
-getSSOStatus :: TeamId -> Handler Bool
+getSSOStatus :: TeamId -> Handler SetSSOStatus
 getSSOStatus tid = do
     info $ msg "Getting SSO status"
     gly <- view galley
     (>>= fromResponseBody) . catchRpcErrors $ rpc' "galley" gly
          ( method GET
          . paths ["/i/teams", toByteString' tid, "features", "sso"]
-         . expect2xx
          )
   where
-    fromResponseBody :: Response (Maybe LByteString) -> Handler Bool
+    fromResponseBody :: Response (Maybe LByteString) -> Handler SetSSOStatus
     fromResponseBody resp = case responseJsonEither resp of
-      Right (SSOTeamConfig SSODisabled) -> pure False
-      Right (SSOTeamConfig SSOEnabled) -> pure True
+      Right (SSOTeamConfig SSOEnabled) -> pure SetSSOEnabled
+      Right (SSOTeamConfig SSODisabled) -> pure SetSSODisabled
       Left errmsg -> throwE (Error status502 "bad-upstream" ("bad response; error message: " <> pack errmsg))
```

Comment:
We have an open issue whereby backoffice returns a 500. [This comment](https://github.com/wearezeta/backend-issues/issues/877#issuecomment-527534209) indicates that this is due to an unhandled 403.That not being handled is in the code path I'm indicating here.

In the current version of deployed stern, this is indeed the case:

```
curl -XPUT localhost:8080/teams/224de122-d4ae-11e9-8e59-00163e5e6c23/features/sso -d 'false' -H "Content-Type: application/json"
```

leads to a 500. This 500 is due to, as already @tiago-loureiro pointed out, due to an unhandled 403.

As you can see, in the logs, a `403` from `galley` is returned:

```
2019-09-11T16:20:28.11020 |srv=stern|1:I,7:request,1:=,42:stern-d42ada9d-6a59-4f9d-a87a-cbfcf2819345,18:Setting SSO status,
2019-09-11T16:20:29.11107 |srv=stern|1:E,7:request,1:=,42:stern-d42ada9d-6a59-4f9d-a87a-cbfcf2819345,6:remote,1:=,6:galley,4:path,1:=,58:/i/teams/224de122-d4ae-11e9-8e59-00163e5e6c23/features/sso,10:Request-Id,1:=,42:stern-d42ada9d-6a59-4f9d-a87a-cbfcf2819345,12:Content-Type,1:=,16:application/json,1745:HttpExceptionRequest Request {   host                 = "127.0.0.1"   port                 = 8005   secure               = False   requestHeaders       = [("Content-Type","application/json"),("Request-Id","stern-d42ada9d-6a59-4f9d-a87a-cbfcf2819345")]   path                 = "/i/teams/224de122-d4ae-11e9-8e59-00163e5e6c23/features/sso"   queryString          = ""   method               = "PUT"   proxy                = Nothing   rawBody              = False   redirectCount        = 10   responseTimeout      = ResponseTimeoutDefault   requestVersion       = HTTP/1.1 }  (StatusCodeException (Response {responseStatus = Status {statusCode = 403, statusMessage = "Forbidden"}, responseVersion = HTTP/1.1, responseHeaders = [("Transfer-Encoding","chunked"),("Date","Wed, 11 Sep 2019 16:20:27 GMT"),("Server","Warp/3.2.25"),("Content-Type","application/json")], responseBody = (), responseCookieJar = CJ {expose = []}, responseClose' = ResponseClose}) "{\"code\":403,\"message\":\"The SSO feature flag is disabled by default.  It can only be enabled once for any team, never disabled.\\n\\nRationale: there are two services in the backend that need to keep their status in sync: galley (teams),\\nand spar (SSO).  Galley keeps track of team features.  If galley creates an idp, the feature flag is\\nchecked.  For authentication, spar avoids this expensive check and assumes that the idp can only have\\nbeen created if the SSO is enabled.  This assumption does not hold any more if the switch is turned off\\nagain, so we do not support this.\\n\\nIt is definitely feasible to change this.  If you have a use case, please contact customer support, or\\nopen an issue on https://github.com/wireapp/wire-server.\",\"label\":\"not-implemented\"}"),
2019-09-11T16:20:29.11136 |srv=stern|1:D,4:code,1:=,3:500,5:label,1:=,8:io-error,7:request,1:=,3:N/A,11:"I/O Error",
```

On the code on develop, this is converted by the `expect2xx` to a 500, which is then shown on backoffice.

With the changes in this PR, you removed the `expect2xx`, and instead convert any responsebody that can't be parsed as a SSOStatus to a 502. Now a 502 will be shown on the backoffice. 

The underlying issue, namely that stern doesn't handle the case where galley returns a 403, remains. This is what I meant to point out.

Reply:
Sorry  :)

This vaguely makes sense.  I'll look into it once more, and come back with a better answer...

Reply:
Actually, I just checked out this branch and ran

```
curl -XPUT localhost:8091/teams/224de122-d4ae-11e9-8e59-00163e5e6c23/features/sso -d '"SetSSODisabled"' -H "Content-Type: application/json" -v
```
and I get a 200. Now I'm confused, as I don't understand why this suddenly works now.

Reply:
Hm, okay, there's a 200, but the behaviour is unexpected still:

```
curl -XPUT localhost:8091/teams/224de122-d4ae-11e9-8e59-00163e5e6c23/features/sso -d '"SetSSODisabled"' -H "Content-Type: application/json" -v
```

```
curl  localhost:8091/teams/224de122-d4ae-11e9-8e59-00163e5e6c23/features/sso
"SetSSOEnabled"
```

So it returns 200 but it hasn't applied the change. 


```
# initially
curl  localhost:8091/teams/9707aad6-d4b6-11e9-9f9f-00163e5e6c23/features/sso
"SetSSODisabled"

# enable
curl -XPUT localhost:8091/teams/9707aad6-d4b6-11e9-9f9f-00163e5e6c23/features/sso -d '"SetSSOEnabled"' -H "Content-Type: application/json" -v
                 
# isEnabled, yes
curl  localhost:8091/teams/9707aad6-d4b6-11e9-9f9f-00163e5e6c23/features/sso
"SetSSOEnabled"

# now disable
curl -XPUT localhost:8091/teams/9707aad6-d4b6-11e9-9f9f-00163e5e6c23/features/sso -d '"SetSSODisabled"' -H "Content-Type: application/json" -v
                  
# now query again
curl  localhost:8091/teams/9707aad6-d4b6-11e9-9f9f-00163e5e6c23/features/sso
"SetSSOEnabled"
```

Hm, perhaps disabling an already-enabled sso should not return a 200 then, since it's not taken into account?